R/n_clusters.R

```diff
@@ -0,0 +1,120 @@
+  if("nbclust" %in% tolower(package)){
+    out <- rbind(out, .n_clusters_NbClust(x))
+  }
+  if("mclust" %in% tolower(package)){
+    out <- rbind(out, .n_clusters_mclust(x))
+  }
+  if("clvalid" %in% tolower(package)){
+    out <- rbind(out, .n_clusters_clValid(x))
+  }
+
+
+  out <- out[order(out$n_Clusters), ] # Arrange by n clusters
+  row.names(out) <- NULL # Reset row index
+  out$Method <- as.character(out$Method)
+
+  # Remove duplicate methods starting with the smallest
+  dupli <- c()
+  for(i in 1:nrow(out)){
+    if(i > 1 && out[i, "Method"] %in% out$Method[1:i-1]){
+      dupli <- c(dupli, i)
+    }
+  }
+  out <- out[-dupli, ]
+
+
+  # Add summary
+  by_clusters <- .data_frame(
+    n_Clusters = as.numeric(unique(out$n_Clusters)),
+    n_Methods = as.numeric(by(out, as.factor(out$n_Clusters), function(out) n <- nrow(out)))
+  )
+
+  attr(out, "summary") <- by_clusters
+  attr(out, "n") <- min(as.numeric(as.character(by_clusters[by_clusters$n_Methods == max(by_clusters$n_Methods), c("n_Clusters")])))
+
+  class(out) <- c("n_clusters", "see_n_clusters", class(out))
+  out
+}
+
+
+
+
+
+
+
+
+
+
+
+#' @keywords internal
+.n_clusters_mclust <- function(x, ...) {
+  if (!requireNamespace("mclust", quietly = TRUE)) {
+    stop("Package 'mclust' required for this function to work. Please install it by running `install.packages('mclust')`.")
+  }
+  mclustBIC <- mclust::mclustBIC # this is needed as it is internally required by the following function
+  BIC <- mclust::mclustBIC(x, verbose = FALSE)
+  out <- data.frame(unclass(BIC))
+  n <- which(out == max(out, na.rm = TRUE), arr.ind = TRUE)[1]
+  data.frame(n_Clusters = n, Method = "Mixture", Package = "mclust")
+}
+
+
+#' @keywords internal
+.n_clusters_clValid <- function(x, ...) {
+  if (!requireNamespace("clValid", quietly = TRUE)) {
+    stop("Package 'clValid' required for this function to work. Please install it by running `install.packages('clValid')`.")
+  }
+
+  row.names(x) <- paste0("row", row.names(x))
+  rez <- clValid::clValid(x,
+    nClust = 2:9,
+    # clMethods = c("hierarchical", "kmeans", "pam", "diana", "fanny", "som", "model", "sota", "clara", "agnes"),
+    clMethods = c("hierarchical", "kmeans", "pam", "diana", "clara", "agnes"),
+    validation = c("internal", "stability")
+  )
+  out <- clValid::optimalScores(rez)
+  data.frame(n_Clusters = out$Clusters, Method = row.names(out), Package = "clValid")
+}
+
+
+#' @importFrom grDevices png dev.off
+#' @keywords internal
+.n_clusters_NbClust <- function(x, ...) {
+  if (!requireNamespace("NbClust", quietly = TRUE)) {
+    stop("Package 'NbClust' required for this function to work. Please install it by running `install.packages('NbClust')`.")
+  }
+
+  # Run the function and suppress output and automatic plotting
+  ff <- tempfile()
+  grDevices::png(filename = ff)
+  junk <- capture.output(n <- NbClust::NbClust(x, min.nc = 2, max.nc = 9, method = "ward.D2", index = "alllong"))
+  grDevices::dev.off()
+  unlink(ff)
```

Comment:
Why do we need grDevices here?

Reply:
otherwise it plots a figure automatically

Reply:
grrr...

Reply:
I know right